qi-test/tests/compiler/rules/full-cycle.rkt

```diff
@@ -46,7 +47,46 @@
                (deforested?
                  (phase1-eval
                   (qi-compile
-                   #'(~>> (filter odd?) values (map sqr)))))))))
+                   #'(~>> (filter odd?) values (map sqr)))))))
+   (test-suite
+    "sandboxed evaluation"
+    (test-not-exn "Plays well with sandboxed evaluation"
+                  ;; This test reproduces the bug and the fix fixes it. Yet,
+                  ;; coverage does not show the lambda in `my-emit-local-step`
+                  ;; as being covered. This could be because the constructed
+                  ;; sandbox evaluator "covering" the code doesn't count as
+                  ;; coverage by the main evaluator running the test?
+                  ;; We address this by putting `my-emit-local-step` in a
+                  ;; submodule, which, by default, are ignored by coverage.
+                  (lambda ()
+                    (let ([eval (parameterize ([sandbox-output 'string]
+                                               [sandbox-error-output 'string]
+                                               [sandbox-memory-limit #f])
+                                  (make-evaluator
+                                   'racket/base
+                                   '(require (for-syntax racket/base)
+                                             ;; necessary to recognize and expand core forms correctly
+                                             qi/flow/extended/expander
+                                             ;; necessary to correctly expand the right-threading form
+                                             qi/flow/extended/forms
+                                             syntax/macro-testing
+                                             racket/list
+                                             qi/flow/core/compiler
+                                             (submod qi/flow/extended/expander invoke))
+
+                                   '(begin-for-syntax
+                                      (require syntax/parse/define
+                                               (for-template qi/flow/core/compiler)
+                                               (for-syntax racket/base))
+
+                                      ;; A macro that expands and compiles surface syntax
+                                      (define-syntax-parse-rule (qi-compile stx)
+                                        (compile-flow
+                                         (expand-flow stx))))))])
```

Comment:
I've just noticed that this `qi-compile` and the other one don't really seem to use `syntax-parse` features: In fact, if I didn't know better, since the `stx` they're getting is really a syntax object, I would say this could be a function?


```
(define (qi-compile stx)
  (compile-flow (expand-flow stx)))
```

Now, it may be that `phase1-eval` only makes sense when the expressions being evaluated are available at syntax time; but the definition would be in `begin-for-syntax` here and in the other use.

---

Said another way, using `S` where you have written `stx`, this is like writing

```
(begin-for-syntax
  #;(require …)
  (define-syntax (qi-compile stx)
    (syntax-parse stx
      [(_ S) #'(compile-flow (expand-flow S))])))
```

Ah, except that `compile-flow` and `expand-flow` are probably macros?

The expansion of the whole thing would be weird, though:
```
(phase1-eval (compile-flow (expand-flow #'sqr)))
```

Is the point of writing `qi-compile` as a macro so that it works as expected with `phase1-eval`? Maybe I'm just confused by the use of `stx` as a pattern variable when I often see it only used for the syntax object given to macro invocations.

Reply:
(This comment is a bit of a mess, but it's late. Hopefully the fact that I'm confused is sufficient to extract a good explanation 😅 .)

Reply:
Well spotted. I've made the change!